File: split/scanint.pas
procedure scanint;
label
  30;
var
  negative: boolean;
  m: integer;
  d: smallnumber;
  vacuous: boolean;
  OKsofar: boolean;
begin
  radix := 0;
  OKsofar := true;
  {441:}
negative := false;
  repeat
    {406:}
repeat
      getxtoken;
    until curcmd <> 10{:406}
;
    if curtok = 3117 then
    begin
      negative := not negative;
      curtok := 3115;
    end;
  until curtok <> 3115{:441}
;
  if curtok = 3168 then{442:}

  begin
    gettoken;
    if curtok < 4095 then
    begin
      curval := curchr;
      if curcmd <= 2 then
        if curcmd = 2 then
          alignstate := alignstate + 1
        else
          alignstate := alignstate - 1;
    end
    else if curtok < 4352 then
      curval := curtok - 4096
    else
      curval := curtok - 4352;
    if curval > 255 then
    begin
      begin
        if interaction = 3 then;
        printnl(262);
        print(698);
      end;
      begin
        helpptr := 2;
        helpline[1] := 699;
        helpline[0] := 700;
      end;
      curval := 48;
      backerror;
    end
    else{443:}

    begin
      getxtoken;
      if curcmd <> 10 then
        backinput;
    end{:443}
;
  end{:442}

  else if (curcmd >= 68) and (curcmd <= 89) then
    scansomethinginternal(0,
      false)
  else{444:}

  begin
    radix := 10;
    m := 214748364;
    if curtok = 3111 then
    begin
      radix := 8;
      m := 268435456;
      getxtoken;
    end
    else if curtok = 3106 then
    begin
      radix := 16;
      m := 134217728;
      getxtoken;
    end;
    vacuous := true;
    curval := 0;
    {445:}
while true do
    begin
      if (curtok < 3120 + radix) and (curtok >= 3120) and (curtok <= 3129) then
        d := curtok - 3120
      else if radix = 16 then
        if (curtok <= 2886) and (curtok >= 2881) then
          d := curtok - 2871
        else if (curtok <= 3142) and (curtok >= 3137) then
          d := curtok - 3127
        else
          goto 30
      else
        goto 30;
      vacuous := false;
      if (curval >= m) and ((curval > m) or (d > 7) or (radix <> 10)) then
      begin
        if OKsofar then
        begin
          begin
            if interaction = 3 then;
            printnl(262);
            print(701);
          end;
          begin
            helpptr := 2;
            helpline[1] := 702;
            helpline[0] := 703;
          end;
          error;
          curval := 2147483647;
          OKsofar := false;
        end;
      end
      else
        curval := curval * radix + d;
      getxtoken;
    end;
    30:
    {:445}
;
    if vacuous then{446:}

    begin
      begin
        if interaction = 3 then;
        printnl(262);
        print(664);
      end;
      begin
        helpptr := 3;
        helpline[2] := 665;
        helpline[1] := 666;
        helpline[0] := 667;
      end;
      backerror;
    end{:446}

    else if curcmd <> 10 then
      backinput;
  end{:444}
;
  if negative then
    curval := -curval;
end;

File: cpp/scanint.cpp
#include "scanint.h"
#include "getxtoken.h"
#include "gettoken.h"
#include "printnl.h"
#include "print.h"
#include "backerror.h"
#include "backinput.h"
#include "scansomethinginternal.h"
#include "error.h"
#include "texte.h"

void scanint(void)
{
	radix = 0;
	bool OKsofar = true;
	bool negative = false;
	do
	{
		do
			getxtoken();
		while (curcmd == spacer);
		if (curtok == other_token+'-')
		{
			negative = !negative;
			curtok = other_token+'+';
		}
	} while (curtok == other_token+'+');
	if (curtok == alpha_token)
	{
		gettoken();
		if (curtok < cs_token_flag)
		{
			curval = curchr;
			if (curcmd == right_brace)
				alignstate++;
			if (curcmd == left_brace)
				alignstate--;
		}
		else 
			curval = curtok-cs_token_flag-(curtok < cs_token_flag+single_base ? active_base : single_base);
		if (curval > 0xFF)
		{
			printnl("! ");
			print("Improper alphabetic constant");
			helpptr = 2;
			helpline[1] = "A one-character control sequence belongs after a ` mark.";
			helpline[0] = "So I'm essentially inserting \\0 here.";
			curval = '0';
			backerror();
		}
		else
		{
			getxtoken();
			if (curcmd != spacer)
				backinput();
		}
	}
	else 
		if (curcmd >= min_internal && curcmd <= max_internal)
			scansomethinginternal(0, false);
		else
		{
			radix = 10;
			auto m = 214748364; //0xC'CC'CC'CC
			if (curtok == octal_token) //other_char+'\''
			{
				radix = 8;
				m = 0x10'00'00'00;
				getxtoken();
			}
			else 
				if (curtok == hex_token) //other_char+'\"'
				{
					radix = 16;
					m = 0x8'00'00'00;
					getxtoken();
				};
			bool vacuous = true;
			curval = 0;
			while (true)
			{
				int d;
				if (curtok < zero_token+radix && curtok >= zero_token && curtok <= zero_token+9)
					d = curtok-zero_token;
				else 
					if (radix == 16)
						if (curtok <= A_token+5 && curtok >= A_token)
							d = curtok-A_token+10;
						else 
							if (curtok <= other_A_token+5 && curtok >= other_A_token)
								d = curtok-other_A_token+10;
							else
								break;
					else
						break;
				vacuous = false;
				if (curval >= m && (curval > m || d > 7 || radix != 10))
				{
					if (OKsofar)
					{
						printnl("! ");
						print("Number too big");
						helpptr = 2;
						helpline[1] = "I can only go up to 2147483647='17777777777=\"7FFFFFFF,";
						helpline[0] = "so I'm using that number instead of yours.";
						error();
						curval = infinity;
						OKsofar = false;
					}
				}
				else
					curval = curval*radix+d;
				getxtoken();
			}
			if (vacuous)
			{
				printnl("! ");
				print("Missing number, treated as zero");
				helpptr = 3;
				helpline[2] = "A number should have been here; I inserted `0'.";
				helpline[1] = "(If you can't figure out why I needed to see a number,";
				helpline[0] = "look up `weird error' in the index to The TeXbook.)";
				backerror();
			}
			else 
				if (curcmd != spacer)
					backinput();
		}
	if (negative)
		curval = -curval;
}
